pkg/trace/metrics/timing/timing.go

```diff
@@ -0,0 +1,153 @@
+// Package timing is used to aggregate timing calls within hotpaths to avoid using
+// repeated statsd calls. The package has a default set that reports at 10 second
+// intervals and can be used directly. If a different behaviour or reporting pattern
+// is desired, a custom Set may be created.
+package timing
+
+import (
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/atomic"
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
+)
+
+// AutoreportInterval specifies the interval at which the default set reports.
+const AutoreportInterval = 10 * time.Second
+
+var (
+	defaultSet Set
+	stopReport = defaultSet.Autoreport(AutoreportInterval)
+)
+
+// Since records the duration for the given metric name as time passed since start.
+// It uses the default set which is reported at 10 second intervals.
+func Since(name string, start time.Time) { defaultSet.Since(name, start) }
+
+// Stop permanently stops the default set from auto-reporting and flushes any remaining
+// metrics. It can be useful to call when the program exits to ensure everything is
+// submitted.
+func Stop() { stopReport() }
+
+// Set represents a set of metrics that can be used for timing. A zero value Set is ready
+// to use. Use Report (or Autoreport) to submit metrics. Set is safe for concurrent use.
+type Set struct {
+	mu sync.RWMutex        // guards c
+	c  map[string]*counter // maps names to their aggregates
+}
+
+// Autoreport enables autoreporting of the Set at the given interval. It returns a
+// cancellation function.
+func (s *Set) Autoreport(interval time.Duration) (cancelFunc func()) {
+	stop := make(chan struct{})
+	go func() {
+		defer close(stop)
+		tick := time.NewTicker(interval)
+		defer tick.Stop()
+		for {
+			select {
+			case <-tick.C:
+				s.Report()
+			case <-stop:
+				s.Report()
+				return
+			}
+		}
+	}()
+	var once sync.Once // avoid panics
+	return func() {
+		once.Do(func() {
+			stop <- struct{}{}
+			<-stop
+		})
+	}
+}
+
+// Since records the duration for the given metric name as *time passed since start*.
+// If name does not exist, as defined by NewSet, it creates it.
+func (s *Set) Since(name string, start time.Time) {
+	ms := time.Since(start) / time.Millisecond
+	s.getCounter(name).add(float64(ms))
+}
+
+// getCounter returns the counter with the given name, initializing any unitialized
+// fields of s.
+func (s *Set) getCounter(name string) *counter {
+	s.mu.RLock()
+	mapReady := s.c != nil
+	s.mu.RUnlock()
+	if !mapReady {
```

Comment:
Couldn't we skip all of this by constructing the map along with the defaultset? `defaultSet Set = Set{c: make(map[string]*counter)}`

Reply:
No, because people might want to use `Set` themselves and they won't be able to initialize the private field `c`, nor do we want to give external access to it.